execution_engine/src/storage/trie_store/lmdb.rs

```diff
@@ -161,3 +172,150 @@ impl<K, V> Store<Digest, Trie<K, V>> for LmdbTrieStore {
 }
 
 impl<K, V> TrieStore<K, V> for LmdbTrieStore {}
+
+pub(crate) type Cache = Arc<Mutex<HashMap<Digest, (bool, Trie<Key, StoredValue>)>>>;
+/// In-memory cached trie store, backed by rocksdb.
+#[derive(Clone)]
+pub struct ScratchCache {
+    pub(crate) cache: Cache,
+    pub(crate) store: Arc<LmdbTrieStore>,
+    pub(crate) env: Arc<LmdbEnvironment>,
+}
+
+/// Cached version of the trie store.
+#[derive(Clone)]
+pub struct ScratchTrieStore {
+    pub(crate) inner: ScratchCache,
+}
+
+impl ScratchTrieStore {
+    /// Creates a new ScratchTrieStore.
+    pub fn new(store: Arc<LmdbTrieStore>, env: Arc<LmdbEnvironment>) -> Self {
+        Self {
+            inner: ScratchCache {
+                store,
+                env,
+                cache: Default::default(),
+            },
+        }
+    }
+
+    /// Writes all dirty (modified) cached tries to underlying db.
+    pub fn write_all_tries_to_db(self, new_state_root: Digest) -> Result<(), error::Error> {
+        let env = self.inner.env;
+        let store = self.inner.store;
+        let cache = &mut *self.inner.cache.lock().map_err(|_| error::Error::Poison)?;
+
+        let mut missing_trie_keys = vec![new_state_root];
+        let mut validated_tries = HashMap::new();
+
+        let mut txn = env.create_read_write_txn()?;
+
+        while let Some(next_trie_key) = missing_trie_keys.pop() {
+            if cache.is_empty() {
```

Comment:
It's not clear to me why we need this `if` block.  This would return an error, even if `next_trie_key` was available in the underlying DB.  Seems like we could delete this block and just let the `None` case of the subsequent `match cache.remove()` handle it.

Reply:
resolved in [4b78471](https://github.com/casper-network/casper-node/pull/2877/commits/4b78471c0fbac1bcaf8999792dcda1d0ded85d70)